component/config.go

```diff
@@ -33,12 +37,43 @@ type ConfigValidator interface {
 // ValidateConfig validates a config, by doing this:
 //   - Call Validate on the config itself if the config implements ConfigValidator.
 func ValidateConfig(cfg Config) error {
-	validator, ok := cfg.(ConfigValidator)
-	if !ok {
-		return nil
+	return validate(reflect.ValueOf(cfg))
+}
+
+func validate(t reflect.Value) error {
+	// Validate the value itself.
+	var errs error
+	// If implements ConfigValidator then call Validate.
+	if validator, ok := t.Interface().(ConfigValidator); ok {
+		errs = multierr.Append(errs, validator.Validate())
+	}
```

Comment:
Maybe we should validate nested fields first? I can imagine that some validation logic relying on already validated nested fields.

Reply:
Actually nvm. Since we don't return and combine all the errors, it only affects the order of the errors.